--- ultramonkey-l7-r10234/l7vsd/module/protocol/protocol_module_ip.cpp	2010-05-07 13:17:28.000000000 +0900
+++ ultramonkey-l7-3.0.0-0/l7vsd/module/protocol/protocol_module_ip.cpp	2010-09-17 17:09:16.188931422 +0900
@@ -56,6 +56,9 @@
 const int protocol_module_ip::FORWARDED_FOR_OFF = 0;
 const int protocol_module_ip::FORWARDED_FOR_ON = 1;
 
+const int protocol_module_ip::COLLECT_STATS_OFF = 0;
+const int protocol_module_ip::COLLECT_STATS_ON = 1;
+
 using namespace boost::xpressive;
 //! constractor
 protocol_module_ip::protocol_module_ip() :
@@ -323,6 +326,7 @@
         bool no_reschedule_flag = false;
         bool forward_checked = false;
         bool sorryuri_checked = false;
+        bool stats_checked = false;
         sregex    sorry_uri_regex
         =    +('/' >>
                *(alpha |
@@ -492,13 +496,61 @@
                                         break;
                                 }
                         }
+                        //option string = "-c/--statistic"
+                        else if (*it == "-c" || *it == "--statistic") {
+                                //statistic flag is OFF
+                                if (!stats_checked) {
+                                        //next item exist
+                                        if(++it != it_end) {
+                                            //collect statistic flag must be 0 or 1
+                                            if(*it == "0" || *it == "1"){
+                                                        //check OK
+                                                        //set statistic flag ON
+                                                        stats_checked = true;                                            }
+                                            else {
+                                                        std::ostringstream ostr;
+                                                        ostr << "'-c/--statistic' option value '" << *it << "' is not a valid value.";
+
+                                                        //set check result flag false
+                                                        check_result.flag = false;
+                                                        //set check result message
+                                                        check_result.message = ostr.str();
+                                                        putLogError(600114, check_result.message, __FILE__, __LINE__);
+                                                        //loop break
+                                                        break;
+                                                }
+                                        }
+                                        //next item is not exist
+                                        else {
+                                                //set check flag false
+                                                check_result.flag = false;
+                                                //set check result message
+                                                check_result.message = "You have to set option value '-c/--statistic'.";
+                                                putLogError(600115, check_result.message, __FILE__,
+                                                            __LINE__);
+                                                //loop break
+                                                break;
+                                        }
+                                }
+                                //statistic flag is ON
+                                else {
+                                        //set check result flag false
+                                        check_result.flag = false;
+                                        //set check result message
+                                        check_result.message = "Cannot set multiple option '-c/--statistic'.";
+                                        putLogError(600116, check_result.message, __FILE__,
+                                                    __LINE__);
+                                        //loop break
+                                        break;
+                                }
+                        }
                         //other option string
                         else {
                                 //set check result flag false
                                 check_result.flag = false;
                                 //set check result message
                                 check_result.message = "Option error.";
-                                putLogError(600010, check_result.message, __FILE__, __LINE__);
+                                putLogError(600117, check_result.message, __FILE__, __LINE__);
                                 //loop break
                                 break;
                         }
@@ -565,6 +617,7 @@
         bool no_reschedule_flag = false;
         bool forward_checked = false;
         bool sorryuri_checked = false;
+        bool stats_checked = false;
         boost::format formatter;
         sregex    sorry_uri_regex
         =    +('/' >>
@@ -731,6 +784,57 @@
                                                 break;
                                         }
                                 }
+                        //option string = "-c/--statistic"
+                        else if (*it == "-c" || *it == "--statistic") {
+                                //statistic flag is OFF
+                                if (!stats_checked) {
+                                        //next item exist
+                                        if(++it != it_end) {
+                                            //collect statistic flag must be 0 or 1
+                                            if(*it == "0" || *it == "1"){
+                                                        //check OK
+                                                        //set statistic flag ON
+                                                        stats_checked = true;
+
+                                                        //set collect statistic flag
+                                                        statistic = boost::lexical_cast<int>(*it);                                          }
+                                            else {
+                                                        std::ostringstream ostr;
+                                                        ostr << "'-c/--statistic' option value '" << *it << "' is not a valid value.";
+
+                                                        //set check result flag false
+                                                        check_result.flag = false;
+                                                        //set check result message
+                                                        check_result.message = ostr.str();
+                                                        putLogError(100002, check_result.message, __FILE__, __LINE__);
+                                                        //loop break
+                                                        break;
+                                                }
+                                        }
+                                        //next item is not exist
+                                        else {
+                                                //set check flag false
+                                                check_result.flag = false;
+                                                //set check result message
+                                                check_result.message = "You have to set option value '-c/--statistic'.";
+                                                putLogError(100003, check_result.message, __FILE__,
+                                                            __LINE__);
+                                                //loop break
+                                                break;
+                                        }
+                                }
+                                //statistic flag is ON
+                                else {
+                                        //set check result flag false
+                                        check_result.flag = false;
+                                        //set check result message
+                                        check_result.message = "Cannot set multiple option '-c/--statistic'.";
+                                        putLogError(100004, check_result.message, __FILE__,
+                                                    __LINE__);
+                                        //loop break
+                                        break;
+                                }
+                        }
                                 //sorryURI flag = ON
                                 else {
                                         //set check result flag false
@@ -770,6 +874,10 @@
                         if (!forward_checked) {
                                 forwarded_for = 0;
                         }
+                        //collect statistic flag = OFF
+                        if (!stats_checked) {
+                                statistic = COLLECT_STATS_OFF;
+                        }
 
                 }
                 /*-------- DEBUG LOG --------*/
@@ -1023,8 +1131,9 @@
         }
         /*------DEBUG LOG END------*/
 
-        boost::format option_formatter("--timeout %d%s %s --sorry-uri '%s'");
-        option_formatter % timeout % (forwarded_for ? " --forwarded-for" : "") % (reschedule ? "--reschedule" : "--no-reschedule") % sorry_uri.c_array();
+        boost::format option_formatter("--timeout %d%s %s --sorry-uri '%s' --statistic '%d'");
+        option_formatter % timeout % (forwarded_for ? " --forwarded-for" : "") % (reschedule ? "--reschedule" : "--no-reschedule")
+                         % sorry_uri.c_array() % statistic;
         option.assign(option_formatter.str());
 
         /*-------- DEBUG LOG --------*/
@@ -1665,6 +1774,17 @@
 
                                                         }
 
+                                                        //increment http statistics
+                                                        increment_stats(session_data_ptr->data_buffer + session_data_ptr->data_offset);
+                                                        /*-------- DEBUG LOG --------*/
+                                                        if (unlikely(LOG_LV_DEBUG == getloglevel())) {
+                                                                boost::format formatter("function : protocol_module_base::EVENT_TAG protocol_module_ip::"
+                                                                                        "handle_client_recv() : call increment_stats : thread id : %d.");
+                                                                formatter % boost::this_thread::get_id();
+                                                                putLogDebug(600228, formatter.str(), __FILE__, __LINE__);
+                                                        }
+                                                        /*------DEBUG LOG END------*/
+
                                                         //set data state HTTP_HEADER
                                                         session_data_ptr->data_state = HTTP_HEADER;
 
@@ -2494,6 +2614,18 @@
 
                                                         }
 
+                                                        //increment http statistics
+                                                        increment_stats(session_data_ptr->data_buffer + session_data_ptr->data_offset);
+                                                        /*-------- DEBUG LOG --------*/
+                                                        if (unlikely(LOG_LV_DEBUG == getloglevel())) {
+                                                                boost::format formatter("function : protocol_module_base::EVENT_TAG protocol_module_ip::"
+                                                                                        "handle_realserver_send() : call increment_stats : thread id : %d.");
+                                                                formatter % boost::this_thread::get_id();
+                                                                putLogDebug(600229, formatter.str(), __FILE__, __LINE__);
+                                                        }
+                                                        /*------DEBUG LOG END------*/
+
+
                                                         //set data state HTTP_HEADER
                                                         session_data_ptr->data_state = HTTP_HEADER;
                                                 }
